ios/NedaaWidget/DatabaseService.swift

```diff
@@ -0,0 +1,150 @@
+import Foundation
+import SQLite3
+
+
+
+
+struct PrayersData: Decodable {
+    let prayerTimes: PrayerTimes
+    let date: String
+    let calculationMethod: Int
+    let timezone: String
+}
+
+struct PrayerTimes: Decodable {
+    let Fajr: String
+    let Sunrise: String
+    let Dhuhr: String
+    let Asr: String
+    let Maghrib: String
+    let Isha: String
+}
+
+struct PrayerData: Decodable {
+    var name: String
+    let date: Date
+}
+
+func parsePrayerData(jsonString: String) -> PrayersData? {
+    let data = jsonString.data(using: .utf8)
+    let decoder = JSONDecoder()
+    do {
+        let prayerData = try decoder.decode(PrayersData.self, from: data!)
+        return prayerData
+    } catch {
+        print("Error decoding JSON: \(error)")
+        return nil
+    }
+}
+
+class DatabaseService {
+    private let appGroupId = "group.io.nedaa.nedaaApp"
+    private let dbPath: String = "nedaa.db"
+    
+    let columnDate = "date"
+    let columnContent = "content"
+    
+    let prayerTimesTable = "PrayerTimes"
+    
+    private var db: OpaquePointer?
+    
+    var timeZone: String = ""
+    
+    
+    func getTimezone()  {
+        let query = "SELECT content from \(prayerTimesTable) LIMIT 1"
+        let queryResult = executeSingleResultQuery(query)
+        let parsedData = parsePrayerData(jsonString: queryResult ?? "")
+        timeZone =  parsedData?.timezone ?? ""
+    }
+    
+    func getDayPrayerTimes(dateInt: Int) -> [PrayerData]? {
+        let query = "SELECT content FROM \(prayerTimesTable) WHERE \(columnDate) = ?"
+        let queryResult = executeSingleResultQuery(query, withParameter: Int32(dateInt))
+        let parsedData = parsePrayerData(jsonString: queryResult ?? "")
+        let prayerTimes = [
+            PrayerData(name: "fajr", date: convertStringToDate(timeString: (parsedData?.prayerTimes.Fajr ?? "") ) ?? Date()),
+            PrayerData(name: "sunrise", date: convertStringToDate(timeString: (parsedData?.prayerTimes.Sunrise ?? "") ) ?? Date()),
+            PrayerData(name: "dhuhr", date: convertStringToDate(timeString: (parsedData?.prayerTimes.Dhuhr ?? "") ) ?? Date()),
+            PrayerData(name: "maghrib", date: convertStringToDate(timeString: (parsedData?.prayerTimes.Maghrib ?? "") ) ?? Date()),
+            PrayerData(name: "asr", date: convertStringToDate(timeString: (parsedData?.prayerTimes.Asr ?? "") ) ?? Date()),
+            PrayerData(name: "isha", date: convertStringToDate(timeString: (parsedData?.prayerTimes.Isha ?? "") ) ?? Date())
+        ]
+        return prayerTimes
+    }
+    
+    private func executeSingleResultQuery(_ query: String, withParameter parameter: Int32? = nil) -> String? {
+        if !openDB() { return nil }
+        // Closing db: not closing it will cause an issue when we open the app
+        // as it will be locked
+        defer { closeDB() }
```

Comment:
In our case. Its better than leaving it open. As, we only query a few times in the day. Also, leaving it open seems to cause an issue when we opening the app, as the db will be in a lock state. Since when we open the app we delete the previous days records from the db.